File: server/src/main/java/org/hkurh/doky/filestorage/DokyLocalFilesystemStorage.java
package org.hkurh.doky.filestorage;

import org.apache.commons.lang3.StringUtils;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.lang.NonNull;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;

import static java.lang.String.format;

@Service
@ConditionalOnProperty(name = "doky.filestorage.type", havingValue = "local-filesystem", matchIfMissing = true)
public class DokyLocalFilesystemStorage implements FileStorage {

    private static final Log LOG = LogFactory.getLog(DokyLocalFilesystemStorage.class);

    private static void saveFileToFilesystem(MultipartFile file, Path path) throws IOException {
        Files.copy(file.getInputStream(), path, StandardCopyOption.REPLACE_EXISTING);
        LOG.debug(format("Save uploaded file to [%s]", path.toAbsolutePath()));
    }

    @Override
    public void saveFile(@NonNull MultipartFile file, @NonNull String filePathWithName) throws IOException {
        var path = Paths.get(filePathWithName);
        saveFileToFilesystem(file, path);
    }

    @Override
    public void saveFile(@NonNull MultipartFile file, @NonNull String filePath, @NonNull String fileName) throws IOException {
        var folder = Paths.get(filePath);
        Files.createDirectories(folder);
        var path = folder.resolve(fileName);
        saveFileToFilesystem(file, path);
    }

    @Override
    public Path getFile(@NonNull String filePath) {
        var file = Paths.get(filePath);
        if (Files.exists(file)) {
            return file;
        } else {
            return null;
        }
    }

    @Override
    public boolean checkExistence(String filePath) {
        if (StringUtils.isBlank(filePath)) {
            return false;
        }
        var file = Paths.get(filePath);
        return Files.exists(file);
    }
}

File: server/src/main/java/org/hkurh/doky/filestorage/DokyLocalFilesystemStorage.kt
package org.hkurh.doky.filestorage

import org.apache.commons.lang3.StringUtils
import org.apache.commons.logging.LogFactory
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.lang.NonNull
import org.springframework.stereotype.Service
import org.springframework.web.multipart.MultipartFile
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.nio.file.StandardCopyOption

@Service
@ConditionalOnProperty(name = ["doky.filestorage.type"], havingValue = "local-filesystem", matchIfMissing = true)
class DokyLocalFilesystemStorage : FileStorage {
    @Throws(IOException::class)
    override fun saveFile(@NonNull file: MultipartFile?, @NonNull filePathWithName: String?) {
        val path = Paths.get(filePathWithName)
        saveFileToFilesystem(file, path)
    }

    @Throws(IOException::class)
    override fun saveFile(@NonNull file: MultipartFile?, @NonNull filePath: String?, @NonNull fileName: String?) {
        val folder = Paths.get(filePath)
        Files.createDirectories(folder)
        val path = folder.resolve(fileName)
        saveFileToFilesystem(file, path)
    }

    override fun getFile(@NonNull filePath: String?): Path? {
        val file = Paths.get(filePath)
        return if (Files.exists(file)) {
            file
        } else {
            null
        }
    }

    override fun checkExistence(filePath: String?): Boolean {
        if (StringUtils.isBlank(filePath)) {
            return false
        }
        val file = Paths.get(filePath)
        return Files.exists(file)
    }

    companion object {
        private val LOG = LogFactory.getLog(DokyLocalFilesystemStorage::class.java)

        @Throws(IOException::class)
        private fun saveFileToFilesystem(file: MultipartFile?, path: Path) {
            Files.copy(file!!.inputStream, path, StandardCopyOption.REPLACE_EXISTING)
            LOG.debug(String.format("Save uploaded file to [%s]", path.toAbsolutePath()))
        }
    }
}
